### PR TITLE
Create Child Views

### PR DESCRIPTION
**Pull request by jonesandcode**
### CHANGES: 10/16 Afternoon
Talked with @jshier and redesigned the view architecture. The tab bar and navigation controller have been removed and we switched to a root view with a view and container view. New screenshot below represents the current build.

### CHANGES: 10/16 Morning
Suggested changes from @jshier made, I also updated the screenshot.

## Reason
Trello Card: [Create Root and Child Views](https://trello.com/c/sp7CfgxM)
To configure the project correctly to show views through state management (loading, loaded, failed, etc).

## What I Did
Created three views, a Root View, Loading View, and Weather View (ViewController) and only show the views upon minimal state change. 

## How I Did It
Created a navigation and tab bar controller. 
Created a Root View, Loading View, and Weather View.
The app starts with the Loading View, and once it receives the data from the API the Weather View is shown. 

This was accomplished by giving the views a Storyboard ID and adding them as a child view of the root view when they need to be shown to the user.

## How You Can Test It
Open the app and run it, and make sure both screens are hit. One is blue and one is turquoise. The weather call loads very quickly so you may have to do it a few times. 

The address leftBarButton needs to be configured to stay the same font size on click. This was not specific to this card, so it will be fixed on it's own card. 

## Screenshots
[Image from Gyazo](https://gyazo.com/9a4d268086681b632822ec99438d093e)

### CODE REVIEW

File: weather/weather/Controllers/LoadViewController.swift [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                self.passWeatherData(weatherModel: weather)
+                self.showWeatherview()
+            } else {
+                print(error!)
+            }
+        }
+    }
+    
+    private func showWeatherview() {
+        if let weatherViewController = storyboard?.instantiateViewController(withIdentifier: "Weather") {
```
Comment (jshier) [inline weather/weather/Controllers/LoadViewController.swift]: Using `guard` for many of these `if let` statements would help with readability.

File: weather/weather/Controllers/LoadViewController.swift [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            addView(weatherViewController)
```
Comment (jshier) [inline weather/weather/Controllers/LoadViewController.swift]: You're embedding the weather controller in the loading controller. Really, you'd watch to switch the loading controller for the weather controller from the root controller, as there really isn't a parent/child relationship between loading and weather.

File: weather/weather/Controllers/LoadViewController.swift [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        super.viewDidLoad()
+        fetchWeather()
+    }
+}
+
+// MARK: ViewControllerDelegate Methods
+
+extension LoadViewController: ViewControllerDelegate {
+    func passWeatherData(weatherModel: WeatherModel) {
+        let viewController = ViewController()
```
Comment (jshier) [inline weather/weather/Controllers/LoadViewController.swift]: Probably want to rename `ViewController` to the relevant name.

File: weather/weather/Models/AddingViewHelper.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+//
+//  Created by Jordan on 10/11/18.
+//  Copyright © 2018 Jordan Jones. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+extension UIViewController {
+    func addView(_ child: UIViewController) {
```
[the remark this answers is not in the document]
Comment (jshier) [reply weather/weather/Models/AddingViewHelper.swift]: Also, this action is usually called "embedding".

File: weather/weather/Models/EmbedView.swift [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+//
+
+import Foundation
+import UIKit
+
+extension UIViewController {
+
+    // Call from parent
+    func embed(_ viewController: UIViewController, in view: UIView) {
+        viewController.willMove(toParent: self)
```
Comment (jshier) [inline weather/weather/Models/EmbedView.swift]: This actually needs to be the `didMove` call at the end.

File: weather/weather/Models/EmbedView.swift [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        addChild(viewController)
+        view.addSubview(viewController.view)
+        viewController.view.frame = view.bounds
```
Comment (jshier) [inline weather/weather/Models/EmbedView.swift]: See what happens when you remove this, other examples don't seem to need it.

File: weather/weather/Models/EmbedView.swift [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    
+    // Call from parent
+    func remove(_ viewController: UIViewController?) {
+        viewController?.removeFromEmbedding()
+    }
+    
+    // Call on children
+    func embedChild(in viewController: UIViewController, view: UIView) {
```
Comment (jshier) [inline weather/weather/Models/EmbedView.swift]: `embedChild` doesn't really makes sense as a name here. `embed(in:view:` should be enough to disambiguate.

File: weather/weather/Controllers/RootViewController.swift [hunk 8]
```diff
@@ -12,11 +12,36 @@ class RootViewController: UIViewController {
...
+        children.first?.removeFromEmbedding()
+        embed(weatherView, in: containerView)
     }
 }
+
+// MARK: Weather Network Request
+
+extension RootViewController {    
+    func getWeather() {
+        networkRequest.fetchWeather(url: "https://api.darksky.net/forecast/\(weatherKey.apiKey)/42.2808256,-83.7430378") { [unowned self] (weather, error) in
```
Comment (jshier) [inline weather/weather/Controllers/RootViewController.swift]: Why the `[unowned self]`?

File: weather/weather/Controllers/RootViewController.swift [hunk 9]
```diff
@@ -12,11 +12,36 @@ class RootViewController: UIViewController {
...
+            guard let weather = weather else { return print("Error: \(error)") }
```
Comment (jshier) [inline weather/weather/Controllers/RootViewController.swift]: Why does `return print...` work?

**Comment by jshier** [comment]
You should update your .gitignore file to get rid of the `.DS_Store` files. GitHub has a default one you can use: https://github.com/github/gitignore/blob/master/Swift.gitignore.

File: weather/weather/Controllers/RootViewController.swift (hunk 9 above)
Comment (jonesandcode) [reply weather/weather/Controllers/RootViewController.swift]: Not really sure about this one?  Does a print statement usually not work in a return statement?

File: weather/weather/Models/EmbedView.swift (hunk 5 above)
Comment (jonesandcode) [reply weather/weather/Models/EmbedView.swift]: Good catch, the changes have been made.

File: weather/weather/Models/EmbedView.swift (hunk 6 above)
Comment (jonesandcode) [reply weather/weather/Models/EmbedView.swift]: Very cool, I tried this and it did work. The "call on parents" methods are commented out. Just to make sure I am understanding correctly, is the code below what you are suggesting?

```    
    // Call on children
    func embed(_ viewController: UIViewController, in view: UIView) {
        addChild(viewController)
        view.addSubview(viewController.view)
        viewController.view.frame = view.bounds
        viewController.didMove(toParent: self)
    }
    
    // Call on children
    func removeEmbeddedView() {
        guard parent != nil else { return }
        
        willMove(toParent: nil)
        removeFromParent()
        view.removeFromSuperview()
    }
```

File: weather/weather/Models/EmbedView.swift (hunk 7 above)
Comment (jonesandcode) [reply weather/weather/Models/EmbedView.swift]: Totally agreed, I didn't notice the "in" difference between the two. Thanks for clarifying!

**Pull request merged**